drivers/sensor/si7006/si7006.c

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (c) 2019 Electronut Labs
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <sensor.h>
+#include <kernel.h>
+#include <device.h>
+#include <init.h>
+#include <string.h>
+#include <sys/byteorder.h>
+#include <sys/__assert.h>
+#include <logging/log.h>
+#include <i2c.h>
+#include <logging/log.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include "si7006.h"
+
+LOG_MODULE_REGISTER(si7006, CONFIG_SENSOR_LOG_LEVEL);
+
+struct si7006_data {
+	struct device *i2c_dev;
+	u16_t temperature;
+	u16_t humidity;
+};
+
+/**
+ * @brief function to get relative humidity
+ *
+ * @return int 0 on success
+ */
+static int si7006_get_humidity(struct device *i2c_dev,
+			       struct si7006_data *si_data)
+{
+	int retval;
+	u8_t hum[2];
+
+	retval = i2c_burst_read(i2c_dev, DT_INST_0_SILABS_SI7006_BASE_ADDRESS,
+		SI7006_MEAS_REL_HUMIDITY_MASTER_MODE, hum, sizeof(hum));
+
+	if (retval == 0) {
+		si_data->humidity = (hum[0] << 8) | hum[1];
+	} else {
+		LOG_ERR("read register err");
+	}
+
+	return retval;
+}
+
+/**
+ * @brief function to get temperature
+ *
+ * @return int 0 on success
+ */
+
+static int si7006_get_temperature(struct device *i2c_dev,
+				  struct si7006_data *si_data)
+{
+	u8_t temp[2];
+	int retval;
+
+	retval = i2c_burst_read(i2c_dev, DT_INST_0_SILABS_SI7006_BASE_ADDRESS,
+		SI7006_MEAS_TEMP_MASTER_MODE, temp, sizeof(temp));
+
+	if (retval == 0) {
+		si_data->temperature = (temp[0] << 8) | temp[1];
+	} else {
+		LOG_ERR("read register err");
+	}
+
+	return retval;
+}
+
+/**
+ * @brief fetch a sample from the sensor
+ *
+ * @return 0
+ */
+static int si7006_sample_fetch(struct device *dev, enum sensor_channel chan)
+{
+	int retval;
+	struct si7006_data *si_data = dev->driver_data;
+
+	retval = si7006_get_temperature(si_data->i2c_dev, si_data);
+	if (retval == 0) {
+		retval = si7006_get_humidity(si_data->i2c_dev, si_data);
+	}
+
+	return retval;
+}
+
+/**
+ * @brief sensor value get
+ *
+ * @return -ENOTSUP for unsupported channels
+ */
+static int si7006_channel_get(struct device *dev, enum sensor_channel chan,
+			      struct sensor_value *val)
+{
+	struct si7006_data *si_data = dev->driver_data;
+
+	if (chan == SENSOR_CHAN_AMBIENT_TEMP) {
+
+		s32_t temp_ucelcius = ((17572 * (s32_t)si_data->temperature)
+				       / 65536) * 10000;
+
+		val->val1 = temp_ucelcius / 1000000;
+		val->val2 = temp_ucelcius % 1000000;
+
+		LOG_DBG("temperature= %d.%06d", val->val1, abs(val->val2));
```

Comment:
This is still a bit imperfect, as for example for both values -0.5 and 0.5 it will print the same: `0.500000`. Since this is only a debug message, maybe it would be easier to just dump `val1` and `val2` directly, without trying to compose a temperature value from them.

Reply:
Agreed.  The existing sensor value representation is fundamentally broken because it doesn't define a normalized representation.  There's no way to reliably print it; @anangl's proposal is a good solution.

Reply:
yes, it makes it easier also. Thanks